Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithInProcessRenderingBackingStore.mm

```diff
@@ -111,19 +112,31 @@
     return setBufferNonVolatile(m_frontBuffer);
 }
 
-RefPtr<WebCore::ImageBuffer> RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer() const
+template<typename ImageBufferType>
+static RefPtr<ImageBuffer> allocateBufferInternal(RemoteLayerBackingStore::Type type, const WebCore::FloatSize& logicalSize, WebCore::RenderingPurpose purpose, float resolutionScale, const WebCore::DestinationColorSpace& colorSpace, WebCore::PixelFormat pixelFormat, WebCore::ImageBufferCreationContext& creationContext)
 {
-    switch (type()) {
-    case RemoteLayerBackingStore::Type::IOSurface: {
-        ImageBufferCreationContext creationContext;
-        creationContext.surfacePool = &WebCore::IOSurfacePool::sharedPool();
-        return WebCore::ImageBuffer::create<ImageBufferShareableMappedIOSurfaceBackend>(size(), scale(), colorSpace(), pixelFormat(), WebCore::RenderingPurpose::LayerBacking, creationContext);
-    }
+    switch (type) {
+    case RemoteLayerBackingStore::Type::IOSurface:
+        return WebCore::ImageBuffer::create<ImageBufferShareableMappedIOSurfaceBackend, ImageBufferType>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext);
     case RemoteLayerBackingStore::Type::Bitmap:
-        return WebCore::ImageBuffer::create<ImageBufferShareableBitmapBackend>(size(), scale(), colorSpace(), pixelFormat(), WebCore::RenderingPurpose::LayerBacking, { });
+        return WebCore::ImageBuffer::create<ImageBufferShareableBitmapBackend, ImageBufferType>(logicalSize, resolutionScale, colorSpace, pixelFormat, purpose, creationContext);
     }
 }
 
+RefPtr<WebCore::ImageBuffer> RemoteLayerWithInProcessRenderingBackingStore::allocateBuffer() const
+{
+    auto purpose = m_layer->containsBitmapOnly() ? WebCore::RenderingPurpose::BitmapOnlyLayerBacking : WebCore::RenderingPurpose::LayerBacking;
+    ImageBufferCreationContext creationContext;
+    creationContext.surfacePool = &WebCore::IOSurfacePool::sharedPool();
+
+#if ENABLE(RE_DYNAMIC_CONTENT_SCALING)
+    if (m_parameters.includeDisplayList == IncludeDisplayList::Yes)
+        return allocateBufferInternal<DynamicContentScalingBifurcatedImageBuffer>(type(), size(), purpose, scale(), colorSpace(), pixelFormat(), creationContext);
+#endif
+
+    return allocateBufferInternal<ImageBuffer>(type(), size(), purpose, scale(), colorSpace(), pixelFormat(), creationContext);
```

Comment:
Don't we need to check `isDynamicContentScalingEnabled()`? Also this code is repeated above. Can't it be shared somehow?

Reply:
`m_parameters.includeDisplayList` already checks `isDynamicContentScalingEnabled`